src/lib/src/grid3d/grid.cpp

```diff
@@ -247,4 +247,66 @@ create_grid_from_cube(const cube::Cube &cube,
     return std::make_tuple(coordsv, zcornsv, actnumsv);
 }
 
+std::tuple<py::array_t<float>, py::array_t<int8_t>>
+refine_vertically(const Grid &grid_cpp, const py::array_t<int8_t> refinement_per_layer)
+{
+    auto actnumsv_ = grid_cpp.actnumsv.unchecked<3>();
+    auto zcornsv_ = grid_cpp.zcornsv.unchecked<4>();
+    auto refinement_data = refinement_per_layer.unchecked<1>();
+
+    // logging here, more for demonstration than anything else
+    auto &logger = xtgeo::logging::LoggerManager::get("xtgeo.grid3d.refine_vertically");
```

Comment:
Perhaps sufficient with `auto &logger = xtgeo::logging::LoggerManager::get("refine_vertically")`, since `xtgeo.grid3d.refine_vertically` is not a correct namespace